src/app_engine/gcm_notify.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/python2.4
+# Copyright 2015 Google Inc. All Rights Reserved.
+"""Module for communication with GCM server.
+"""
+
+import json
+import logging
+import os
+
+import constants
+
+from google.appengine.api import urlfetch
+
+# Load GCM API key from configuration file.
+GCM_API_KEY = None
+GCM_API_URL = 'https://android.googleapis.com/gcm/send'
+
+GCM_MESSAGE_TYPE_BYE = 'BYE'
+GCM_MESSAGE_TYPE_INVITE = 'INVITE'
+GCM_MESSAGE_TYPE_KEY = 'type'
+GCM_MESSAGE_REASON_KEY = 'reason'
+GCM_MESSAGE_REASON_TYPE_DECLINED = 'calleeDeclined'
+GCM_MESSAGE_REASON_TYPE_ACCEPTED = 'calleeAccepted'
+GCM_MESSAGE_REASON_TYPE_HANGUP = 'callerHangup'
+
+
+def get_gcm_api_key():
+  """Reads GCM API key from disk and caches it."""
+  global GCM_API_KEY
+  if GCM_API_KEY is not None:
+    return GCM_API_KEY
+  try:
+    # TODO(tkchin): figure out why unit test can't find this file.
+    with open('gcm_config.json', 'r') as f:
+      gcm_config = json.loads(f.read())
+      gcm_api_key = gcm_config.get('GCM_API_KEY')
+    if not gcm_api_key:
+      logging.error('Failed to load GCM API key from config file.')
+    else:
+      GCM_API_KEY = gcm_api_key
+  except IOError as e:
+    logging.error('Failed to open GCM config file: %s', str(e))
+  return GCM_API_KEY
+
+
+def send_gcm_messages(gcm_ids, message):
+  """Sends the message to each endpoint specified in gcm_ids.
+
+  Makes asynchronous posts to the GCM server with the appropriate payload
+  format. GCM server will then deliver the message the each endpoint. Will
+  block until all posts complete or fail.
+
+  Args:
+    gcm_ids: List of verified GCM ids.
+    message: Dictionary containing custom data.
+  """
+  gcm_api_key = get_gcm_api_key()
+  if gcm_api_key is None:
+    logging.warning('Did not send GCM message due to missing API key.')
+    return
+
+  def handle_rpc_result(rpc):
+    result = rpc.get_result()
+    status_code = result.status_code
+    if status_code != 200:
+      logging.error('Failed to send GCM message. Result:%d', status_code)
+
+  headers = {
+      'Content-Type': 'application/json',
```

Comment:
Still same issue, just that there's a buffer. GCM only supports 4 collapse keys.